File: packages/flow-sdk/src/channels.ts
import { decodeJson, encodeJson } from './json.js'
import {
  parseChannelGrant,
  requireExactKeys,
  requireLocalName,
  requireObject,
  requireWireId,
  type ChannelGrant,
} from './protocol.js'
import {
  OperationError,
  OPERATION_ERROR_CODES,
  type CallOptions,
  type ChannelBroadcast,
  type ChannelEndpoint,
  type ChannelOptions,
  type ChannelPair,
  type ChannelReceiver,
  type ChannelSender,
  type JsonObject,
  type JsonValue,
} from './types.js'

export type ChannelMethod =
  | 'channel/create'
  | 'channel/subscribe'
  | 'channel/send'
  | 'channel/next'
  | 'channel/close'
  | 'channel/release'
export type Settlement = { readonly result: JsonValue } | { readonly error: unknown }
export interface RequestHooks {
  readonly control?: boolean
  readonly settled?: (settlement: Settlement, exposed: boolean, fromWire: boolean) => void
  readonly cancelled?: () => void
}
type Request = (
  method: ChannelMethod,
  params: JsonObject,
  options?: CallOptions,
  hooks?: RequestHooks,
) => Promise<JsonValue>

interface EndpointState {
  readonly grant: ChannelGrant
  readonly endpoint: ChannelEndpoint
  pair?: EndpointState
  used: boolean
  connected: boolean
  offered: boolean
  closed: boolean
  iterator: boolean
  sends: number
  lastSequence: number
  endSequence?: number
  releasedEndSequence?: number
  read?: Promise<void>
  seal?: Promise<void>
  disposal?: Promise<void>
  cause?: unknown
  exposed: boolean
}

/** Process-local endpoint bookkeeping; the host alone commits transfer rights. */
export class Channels {
  private readonly states = new Map<string, EndpointState>()
  private readonly sources = new Set<string>()
  private readonly brands = new WeakMap<ChannelEndpoint, EndpointState>()
  private readonly background = new Set<Promise<unknown>>()
  private allocationCleanupFailed = false

  constructor(private readonly request: Request) {}

  get size(): number {
    return this.states.size
  }

  incoming(
    grants: Readonly<Record<string, ChannelGrant>>,
  ): Readonly<Record<string, ChannelEndpoint>> {
    const result: Record<string, ChannelEndpoint> = Object.create(null)
    for (const [name, grant] of Object.entries(grants)) {
      const state = this.register(grant)
      // Inherited receivers are connected even before their first read.
      state.connected = grant.direction === 'receive'
      result[name] = state.endpoint
    }
    return Object.freeze(result)
  }

  mappings(value: Readonly<Record<string, ChannelEndpoint>> | undefined): JsonObject | undefined {
    if (value === undefined) return undefined
    if (value === null || typeof value !== 'object' || Array.isArray(value))
      throw new TypeError('channels must be a map')
    if (Object.keys(value).length > 256) throw new TypeError('too many channel mappings')
    const result: Record<string, JsonValue> = Object.create(null)
    const states: EndpointState[] = []
    for (const [name, endpoint] of Object.entries(value)) {
      requireLocalName(name)
      const state = this.brands.get(endpoint)
      if (!state || state.used || state.closed)
        throw new TypeError('channels require owned unused endpoints')
      if (states.includes(state)) throw new TypeError('one endpoint cannot fill multiple channels')
      states.push(state)
      result[name] = state.grant.endpoint
    }
    // Offered does not mean moved. Rejected admission leaves rights at the host;
    // joins and explicit later endpoint operations remain host-authoritative.
    for (const state of states) state.offered = true
    return result
  }

  create(
    options: ChannelOptions & { readonly delivery: 'broadcast' },
    callOptions?: CallOptions,
  ): Promise<ChannelBroadcast>
  create(
    options?: ChannelOptions & { readonly delivery?: 'direct' },
    callOptions?: CallOptions,
  ): Promise<ChannelPair>
  create(
    options: ChannelOptions,
    callOptions?: CallOptions,
  ): Promise<ChannelPair | ChannelBroadcast>
  async create(
    options: ChannelOptions = {},
    callOptions?: CallOptions,
  ): Promise<ChannelPair | ChannelBroadcast> {
    if (options === null || typeof options !== 'object' || Array.isArray(options))
      throw new TypeError('channel options must be an object')
    const params = decodeJson(encodeJson(options as unknown as JsonValue)) as JsonObject
    if (Object.keys(params).some((key) => !['delivery', 'schema', 'contract'].includes(key)))
      throw new TypeError('unknown channel option')
    if (
      params.delivery !== undefined &&
      params.delivery !== 'direct' &&
      params.delivery !== 'broadcast'
    )
      throw new TypeError('unsupported channel delivery')
    if (Object.hasOwn(params, 'schema') && Object.hasOwn(params, 'contract'))
      throw new TypeError('schema and contract are exclusive')
    if (
      params.contract !== undefined &&
      (typeof params.contract !== 'string' || !params.contract.startsWith('./'))
    )
      throw new TypeError('contract must be package-local')
    let channel: ChannelPair | ChannelBroadcast | undefined
    await this.request('channel/create', params, callOptions, {
      settled: (settlement, exposed) => {
        if ('error' in settlement) return
        const raw = requireObject(settlement.result, 'channel/create result')
        if (params.delivery === 'broadcast') {
          requireExactKeys(raw, ['send', 'source'])
          const sendGrant = parseChannelGrant(raw.send as JsonValue)
          const source = requireWireId(raw.source as JsonValue)
          if (sendGrant.direction !== 'send' || sendGrant.delivery !== 'broadcast')
            throw new Error('invalid broadcast writer')
          if (this.sources.has(source)) throw new Error('host reused channel source')
          this.sources.add(source)
          const sender = this.register(sendGrant)
          channel = Object.freeze({
            send: sender.endpoint as ChannelSender,
            subscribe: (options?: CallOptions) => this.subscribe(source, sendGrant, options),
          })
          // This writer was never exposed or offered: explicit allocation
          // cleanup is safe, unlike guessing who holds an offered endpoint.
          if (!exposed) this.cleanupAllocation(this.seal(sender))
          return
        }
        requireExactKeys(raw, ['send', 'receive'])
        const sendGrant = parseChannelGrant(raw.send as JsonValue)
        const receiveGrant = parseChannelGrant(raw.receive as JsonValue)
        if (
          sendGrant.direction !== 'send' ||
          receiveGrant.direction !== 'receive' ||
          sendGrant.delivery !== 'direct' ||
          receiveGrant.delivery !== 'direct' ||
          sendGrant.endpoint === receiveGrant.endpoint
        )
          throw new Error('invalid channel pair')
        if (JSON.stringify(sendGrant.contract) !== JSON.stringify(receiveGrant.contract))
          throw new Error('channel pair contract mismatch')
        const sender = this.register(sendGrant)
        const receiver = this.register(receiveGrant)
        sender.pair = receiver
        receiver.pair = sender
        channel = Object.freeze({
          send: sender.endpoint as ChannelSender,
          receive: receiver.endpoint as ChannelReceiver,
        })
        if (!exposed) this.cleanupAllocation(this.dispose(receiver))
      },
    })
    if (!channel) throw new Error('channel allocation did not return endpoints')
    return channel
  }

  private async subscribe(
    source: string,
    writer: ChannelGrant,
    options?: CallOptions,
  ): Promise<ChannelReceiver> {
    let receiver: ChannelReceiver | undefined
    await this.request('channel/subscribe', { source }, options, {
      settled: (settlement, exposed) => {
        if ('error' in settlement) return
        const grant = parseChannelGrant(settlement.result)
        if (grant.direction !== 'receive' || grant.delivery !== 'broadcast')
          throw new Error('invalid broadcast subscription')
        if (JSON.stringify(grant.contract) !== JSON.stringify(writer.contract))
          throw new Error('broadcast subscription contract mismatch')
        const state = this.register(grant)
        // Allocation commits a live subscription even before its first read.
        state.connected = true
        receiver = state.endpoint as ChannelReceiver
        if (!exposed) this.cleanupAllocation(this.dispose(state))
      },
    })
    if (!receiver) throw new Error('subscription allocation did not return a receiver')
    return receiver
  }

  abandoned(): string | undefined {
    for (const state of this.states.values()) {
      if (
        state.grant.direction === 'receive' &&
        !state.offered &&
        !state.closed &&
        !state.disposal &&
        (state.connected || state.used || state.pair?.used || state.pair?.offered)
      )
        return state.grant.endpoint
    }
    return undefined
  }

  async settle(): Promise<void> {
    while (this.background.size) await Promise.allSettled([...this.background])
  }

  async finish(): Promise<void> {
    // Host terminal processing owns offered rights. Never close a token merely
    // because the associated invocation returned success or failure.
    for (const state of this.states.values()) {
      if (state.offered) continue
      if (state.grant.direction === 'receive' && !state.closed) this.track(this.dispose(state))
    }
    await this.settle()
    for (const state of this.states.values()) {
      if (state.grant.direction === 'receive' && !state.offered && state.disposal && !state.closed)
        throw new OperationError('UNCERTAIN', 'channel receiver disposal did not settle')
    }
    if (this.allocationCleanupFailed)
      throw new OperationError('UNCERTAIN', 'cancelled channel allocation cleanup failed')
    // Implicit writer completion belongs to the host's authoritative terminal
    // decision. A caught send error does not reveal whether the source failed.
    // Only an explicit author close issues channel/close from this SDK.
  }

  private register(grant: ChannelGrant): EndpointState {
    if (this.states.has(grant.endpoint)) throw new Error('host reused channel endpoint')
    // A bounded protocol map is the SDK ceiling; hosts may impose lower caps.
    if (this.states.size >= 512) throw new Error('host exceeded endpoint bookkeeping capacity')
    let state: EndpointState
    const common = {
      direction: grant.direction,
      delivery: grant.delivery,
      ...(grant.contract ? { contract: grant.contract } : {}),
    }
    const endpoint: ChannelEndpoint =
      grant.direction === 'send'
        ? Object.freeze({
            ...common,
            direction: 'send' as const,
            send: (value: JsonValue, options?: CallOptions) =>
              this.sendValue(state, value, options),
            close: async (options?: CallOptions) => {
              this.checkSignal(options)
              await this.wait(this.seal(state), options)
            },
          })
        : Object.freeze({
            ...common,
            direction: 'receive' as const,
            startSequence: grant.startSequence!,
            next: (options?: CallOptions) => this.next(state, options),
            close: (options?: CallOptions) => this.closeReceiver(state, options),
            return: async () => {
              await this.closeReceiver(state)
              return { done: true as const, value: undefined }
            },
            [Symbol.asyncIterator]: () => {
              if (state.iterator) throw new TypeError('channel receiver has one iterator')
              state.iterator = true
              state.used = true
              return endpoint as ChannelReceiver
            },
          })
    state = {
      grant,
      endpoint,
      used: false,
      connected: false,
      offered: false,
      closed: false,
      iterator: false,
      sends: 0,
      lastSequence: (grant.startSequence ?? 1) - 1,
      exposed: false,
    }
    this.states.set(grant.endpoint, state)
    this.brands.set(endpoint, state)
    return state
  }

  private async sendValue(
    state: EndpointState,
    value: JsonValue,
    options?: CallOptions,
  ): Promise<void> {
    if (state.closed || state.seal)
      throw new OperationError('DISCONNECTED', 'channel writer is closed')
    const params = decodeJson(encodeJson({ endpoint: state.grant.endpoint, value })) as JsonObject
    state.used = true
    state.sends += 1
    await this.request('channel/send', params, options, {
      settled: (settlement) => {
        state.sends -= 1
        if ('result' in settlement && settlement.result !== null)
          throw new Error('invalid channel/send result')
      },
    })
  }

  private seal(state: EndpointState): Promise<void> {
    if (state.seal) return state.seal
    if (state.closed) return Promise.resolve()
    if (state.sends)
      return Promise.reject(
        new OperationError('INVALID_INPUT', 'channel writer has unaccepted sends'),
      )
    state.used = true
    const promise = this.request('channel/close', { endpoint: state.grant.endpoint }, undefined, {
      control: true,
      settled: (settlement) => {
        if ('result' in settlement) {
          if (settlement.result !== null) throw new Error('invalid channel/close result')
          state.closed = true
        }
      },
    }).then(() => undefined)
    state.seal = promise
    this.track(promise)
    return promise
  }

  private async next(
    state: EndpointState,
    options?: CallOptions,
  ): Promise<IteratorResult<JsonValue>> {
    this.checkSignal(options)
    if (state.read)
      throw new OperationError('INVALID_INPUT', 'channel receiver already has a pending read')
    if (state.cause !== undefined) {
      state.exposed = true
      throw state.cause
    }
    if (state.closed || state.disposal) return { done: true, value: undefined }
    state.used = true
    let resolveRead!: () => void
    state.read = new Promise<void>((resolve) => {
      resolveRead = resolve
    })
    let item: IteratorResult<JsonValue> | undefined
    const promise = this.request('channel/next', { endpoint: state.grant.endpoint }, options, {
      cancelled: () => {
        this.track(this.dispose(state))
      },
      settled: (settlement, exposed, fromWire) => {
        try {
          if ('error' in settlement) {
            if (!fromWire) return
            if (
              !(
                settlement.error instanceof OperationError &&
                settlement.error.code === 'CANCELLED' &&
                state.disposal
              )
            ) {
              state.cause ??= settlement.error
              state.closed = true
              if (exposed) state.exposed = true
            }
            return
          }
          const result = requireObject(settlement.result, 'channel/next result')
          if (Object.hasOwn(result, 'item')) {
            requireExactKeys(result, ['item'])
            const data = requireObject(result.item as JsonValue, 'channel item')
            requireExactKeys(data, ['sequence', 'value'])
            if (data.sequence !== state.lastSequence + 1)
              throw new Error('non-contiguous channel sequence')
            if (
              state.releasedEndSequence !== undefined &&
              data.sequence > state.releasedEndSequence
            )
              throw new Error('channel item exceeds the released end sequence')
            state.lastSequence += 1
            item = { done: false, value: data.value as JsonValue }
          } else {
            requireExactKeys(result, ['end'])
            const end = requireObject(result.end as JsonValue, 'channel end')
            requireExactKeys(end, ['lastSequence'])
            if (end.lastSequence !== state.lastSequence)
              throw new Error('invalid channel end sequence')
            if (
              state.releasedEndSequence !== undefined &&
              state.releasedEndSequence !== state.lastSequence
            )
              throw new Error('channel release disagrees with delivered end')
            state.endSequence = state.lastSequence
            state.closed = true
            item = { done: true, value: undefined }
          }
        } finally {
          delete state.read
          resolveRead()
        }
      },
    })
    await promise
    return item!
  }

  private dispose(state: EndpointState): Promise<void> {
    if (state.disposal) return state.disposal
    const earlierRead = state.read
    state.disposal = this.request(
      'channel/release',
      { endpoint: state.grant.endpoint },
      undefined,
      {
        control: true,
        settled: (settlement) => {
          if ('error' in settlement) return
          const result = requireObject(settlement.result, 'channel/release result')
          if (result.status === 'released') requireExactKeys(result, ['status'])
          else if (result.status === 'ended') {
            requireExactKeys(result, ['status', 'lastSequence'])
            if (
              typeof result.lastSequence !== 'number' ||
              !Number.isSafeInteger(result.lastSequence) ||
              result.lastSequence < state.lastSequence
            )
              throw new Error('invalid released end sequence')
            if (state.endSequence !== undefined && result.lastSequence !== state.endSequence)
              throw new Error('channel release disagrees with delivered end')
            state.releasedEndSequence = result.lastSequence
          } else if (result.status === 'failed') {
            requireExactKeys(result, [
              'status',
              'code',
              ...(Object.hasOwn(result, 'details') ? ['details'] : []),
            ])
            if (
              typeof result.code !== 'string' ||
              !OPERATION_ERROR_CODES.some(
                (code) =>
                  code === result.code && code !== 'CHANNEL_LOST' && code !== 'PROTOCOL_ERROR',
              )
            )
              throw new Error('invalid released channel failure')
            state.cause ??= new OperationError(
              result.code as OperationError['code'],
              'channel delivery failed during disposal',
              result.details,
            )
          } else throw new Error('invalid channel release status')
          state.closed = true
        },
      },
    ).then(async () => {
      await earlierRead
    })
    this.track(state.disposal)
    return state.disposal
  }

  private async closeReceiver(state: EndpointState, options?: CallOptions): Promise<void> {
    this.checkSignal(options)
    await this.wait(this.dispose(state), options)
    if (state.cause !== undefined && !state.exposed) {
      state.exposed = true
      throw state.cause
    }
  }

  private wait<T>(promise: Promise<T>, options?: CallOptions): Promise<T> {
    const signal = options?.signal
    if (!signal) return promise
    if (signal.aborted)
      return Promise.reject(new OperationError('CANCELLED', 'operation wait was cancelled'))
    return new Promise<T>((resolve, reject) => {
      const abort = () => {
        signal.removeEventListener('abort', abort)
        reject(new OperationError('CANCELLED', 'operation wait was cancelled'))
      }
      signal.addEventListener('abort', abort, { once: true })
      promise.then(
        (value) => {
          signal.removeEventListener('abort', abort)
          resolve(value)
        },
        (error) => {
          signal.removeEventListener('abort', abort)
          reject(error)
        },
      )
    })
  }

  private track<T>(promise: Promise<T>): void {
    this.background.add(promise)
    void promise.finally(() => this.background.delete(promise)).catch(() => undefined)
  }

  private checkSignal(options?: CallOptions): void {
    if (options?.signal?.aborted)
      throw new OperationError('CANCELLED', 'operation wait was cancelled')
  }

  private cleanupAllocation(promise: Promise<void>): void {
    this.track(
      promise.catch(() => {
        this.allocationCleanupFailed = true
      }),
    )
  }
}

File: packages/jiggy-flow/src/jiggy/flow/_channels.py
"""Private channel endpoints; authority remains with the Run/1 host."""

from __future__ import annotations

import asyncio
import re
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, Literal, cast

from ._json import MAX_SAFE_INTEGER, normalize_json1
from ._types import ChannelContractIdentity, JsonValue, OperationError

if TYPE_CHECKING:
    from ._runtime import _Runtime


def _object(value: Any, fields: set[str], required: set[str]) -> dict[str, Any]:
    if not isinstance(value, dict) or not required <= value.keys() <= fields:
        raise ValueError("Invalid channel response members")
    return value


def _sequence(value: Any, minimum: int = 0) -> int:
    if type(value) is not int or not minimum <= value <= MAX_SAFE_INTEGER:
        raise ValueError("Invalid channel sequence")
    return value


def validate_grant(value: Any) -> dict[str, Any]:
    grant = _object(value, {"endpoint", "direction", "delivery", "contract", "startSequence"},
                    {"endpoint", "direction", "delivery"})
    validate_reference(grant["endpoint"])
    if (grant["direction"] not in ("send", "receive")
        or grant["delivery"] not in ("direct", "broadcast")):
        raise ValueError("Unsupported channel grant")
    if grant["direction"] == "receive":
        start = _sequence(grant.get("startSequence"), 1)
        if grant["delivery"] == "direct" and start != 1:
            raise ValueError("Direct channel must start at sequence one")
    elif "startSequence" in grant:
        raise ValueError("Sender grant contains receive sequence")
    if "contract" in grant:
        contract = _object(grant["contract"], {"id", "version", "digest"},
                           {"id", "version", "digest"})
        if not isinstance(contract["id"], str) or not _contract_id(contract["id"]):
            raise ValueError("Invalid channel contract ID")
        if not isinstance(contract["version"], str) or re.fullmatch(
            r"(?:0|[1-9][0-9]*)\.(?:0|[1-9][0-9]*)\.(?:0|[1-9][0-9]*)", contract["version"]
        ) is None:
            raise ValueError("Invalid channel contract version")
        if not isinstance(contract["digest"], str) or re.fullmatch(
            r"sha256:[0-9a-f]{64}", contract["digest"]
        ) is None:
            raise ValueError("Invalid channel contract digest")
    return grant


def validate_reference(value: Any) -> str:
    if (not isinstance(value, str) or not 1 <= len(value) <= 128
        or re.fullmatch(r"[A-Za-z0-9][A-Za-z0-9._:/-]*", value) is None):
        raise ValueError("Invalid channel reference")
    return value


def _contract_id(value: str) -> bool:
    match = re.fullmatch(r"https://([^/]+)/(.+)", value)
    if match is None:
        return False
    labels = match[1].split(".")
    return (len(labels) >= 2
            and all(re.fullmatch(r"[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?", label) for label in labels)
            and not (len(labels) == 4
                     and all(re.fullmatch(r"[0-9]{1,3}", label) and int(label) <= 255
                             for label in labels))
            and all(re.fullmatch(r"[a-z0-9._~-]+", part) and part not in (".", "..")
                    for part in match[2].split("/")))


def validate_channel_result(kind: str, value: Any) -> Any:
    if kind in ("channel/send", "channel/close"):
        if value is not None:
            raise ValueError("Channel send/close must return null")
    elif kind == "channel/create":
        created = _object(value, {"send", "receive", "source"}, {"send"})
        send = validate_grant(created["send"])
        if send["direction"] != "send":
            raise ValueError("Channel creation needs a writer")
        if send["delivery"] == "broadcast":
            _object(created, {"send", "source"}, {"send", "source"})
            validate_reference(created["source"])
        else:
            _object(created, {"send", "receive"}, {"send", "receive"})
            receive = validate_grant(created["receive"])
            if (receive["direction"] != "receive" or receive["delivery"] != "direct"
                or send["endpoint"] == receive["endpoint"]
                or send.get("contract") != receive.get("contract")):
                raise ValueError("Invalid direct-channel pair")
    elif kind == "channel/subscribe":
        receive = validate_grant(value)
        if receive["direction"] != "receive" or receive["delivery"] != "broadcast":
            raise ValueError("Subscription needs a broadcast receiver")
    elif kind == "channel/next":
        record = _object(value, {"item", "end"}, set())
        if len(record) != 1:
            raise ValueError("Channel read needs exactly item or end")
        if "item" in record:
            item = _object(record["item"], {"sequence", "value"}, {"sequence", "value"})
            _sequence(item["sequence"], 1)
            normalize_json1(item["value"])
        else:
            end = _object(record["end"], {"lastSequence"}, {"lastSequence"})
            _sequence(end["lastSequence"])
    elif kind == "channel/release":
        if not isinstance(value, dict):
            raise ValueError("Invalid channel release result")
        status = value.get("status")
        if status == "released":
            _object(value, {"status"}, {"status"})
        elif status == "ended":
            _object(value, {"status", "lastSequence"}, {"status", "lastSequence"})
            _sequence(value["lastSequence"])
        elif status == "failed":
            _object(value, {"status", "code", "details"}, {"status", "code"})
            # Use the same operational code vocabulary as ordinary call errors.
            from ._runtime import _FLOW_ERROR_CODES
            if not isinstance(value["code"], str) or value["code"] not in _FLOW_ERROR_CODES:
                raise ValueError("Invalid channel release failure")
        else:
            raise ValueError("Invalid channel release status")
    else:
        raise ValueError("Unknown channel result kind")
    return value


class _Endpoint:
    def __init__(self, runtime: _Runtime, grant: dict[str, Any]):
        self._runtime = runtime
        self._reference: str = grant["endpoint"]
        self._contract: ChannelContractIdentity | None = grant.get("contract")
        self._delivery: Literal["direct", "broadcast"] = grant["delivery"]
        self._used = False
        # An offer is not proof of transfer. The host retains authoritative
        # rights through rejected admission, joins and failed callees.
        self._offered = False

    @property
    def delivery(self) -> Literal["direct", "broadcast"]:
        return self._delivery

    @property
    def contract(self) -> ChannelContractIdentity | None:
        return None if self._contract is None else dict(self._contract)  # type: ignore[return-value]


class _Sender(_Endpoint):
    def __init__(self, runtime: _Runtime, grant: dict[str, Any]):
        super().__init__(runtime, grant)
        self._sealed = False
        self._sends = 0
        self._close_task: asyncio.Task[Any] | None = None

    @property
    def direction(self) -> Literal["send"]:
        return "send"

    async def send(self, value: JsonValue) -> None:
        if self._sealed or self._close_task is not None:
            raise OperationError("DISCONNECTED", "Channel writer is closed")
        value = normalize_json1(value)
        self._used = True
        self._sends += 1
        def settled(*_args: Any) -> None:
            self._sends -= 1
        await self._runtime._send_request(
            "channel/send", "channel/send", {"endpoint": self._reference, "value": value},
            on_settle=settled, on_admission_failure=settled,
        )

    async def close(self) -> None:
        self._used = True
        if self._sealed:
            return
        if self._sends:
            raise OperationError("INVALID_INPUT", "Channel writer has unsettled sends")
        if self._close_task is None:
            async def seal() -> None:
                await self._runtime._send_request("channel/close", "channel/close",
                                                 {"endpoint": self._reference}, settlement=True)
                self._sealed = True
            self._close_task = self._runtime._track_settlement(seal())
        await asyncio.shield(self._close_task)


class _Receiver(_Endpoint):
    def __init__(self, runtime: _Runtime, grant: dict[str, Any]):
        super().__init__(runtime, grant)
        self._start_sequence: int = grant["startSequence"]
        self._last_sequence = self._start_sequence - 1
        self._iterated = False
        self._ended = False
        self._disposed = False
        self._released = False
        self._cause: OperationError | None = None
        self._exposed = False
        self._read_task: asyncio.Task[Any] | None = None
        self._read_settled = asyncio.Event()
        self._read_settled.set()
        self._release_task: asyncio.Task[Any] | None = None

    @property
    def direction(self) -> Literal["receive"]:
        return "receive"

    @property
    def start_sequence(self) -> int:
        return self._start_sequence

    def __aiter__(self) -> _Receiver:
        if self._iterated:
            raise OperationError("OPERATION_CONFLICT", "Channel has one iterator")
        self._iterated = True
        self._used = True
        return self

    def _record_cause(self, cause: OperationError) -> None:
        if self._cause is None and not (self._disposed and cause.code == "CANCELLED"):
            self._cause = cause

    def _settle_read(self, result: Any, is_error: bool) -> None:
        try:
            if is_error:
                self._record_cause(result)
            elif "item" in result:
                sequence = result["item"]["sequence"]
                if sequence != self._last_sequence + 1:
                    raise ValueError("Noncontiguous channel sequence")
                self._last_sequence = sequence
            else:
                if result["end"]["lastSequence"] != self._last_sequence:
                    raise ValueError("Channel end disagrees with delivered prefix")
                self._ended = True
        finally:
            self._read_settled.set()

    async def __anext__(self) -> JsonValue:
        if self._cause is not None:
            self._exposed = True
            raise self._cause
        if self._disposed or self._ended:
            raise StopAsyncIteration
        if self._read_task is not None or not self._read_settled.is_set():
            raise OperationError("OPERATION_CONFLICT", "Channel already has a pending read")
        self._used = True
        self._read_settled.clear()
        self._read_task = asyncio.current_task()
        try:
            result = await self._runtime._send_request(
                "channel/next", "channel/next", {"endpoint": self._reference},
                on_settle=self._settle_read,
                on_admission_failure=self._read_settled.set,
            )
            if "end" in result:
                raise StopAsyncIteration
            return cast(JsonValue, result["item"]["value"])
        except asyncio.CancelledError:
            self._begin_release()
            raise
        except OperationError as error:
            if error is self._cause:
                self._exposed = True
            raise
        finally:
            self._read_task = None

    def _begin_release(self) -> asyncio.Task[Any]:
        self._used = True
        self._disposed = True
        if self._release_task is None:
            if self._read_task is not None and self._read_task is not asyncio.current_task():
                self._read_task.cancel()

            async def dispose() -> None:
                result = await self._runtime._send_request(
                    "channel/release", "channel/release", {"endpoint": self._reference},
                    settlement=True,
                )
                if (result["status"] == "ended"
                    and result["lastSequence"] < self._start_sequence - 1):
                    await self._runtime._fatal_close("PROTOCOL_ERROR")
                    raise OperationError("PROTOCOL_ERROR", "Channel release precedes its subscription")
                self._released = True
                if result["status"] == "failed":
                    self._record_cause(OperationError(result["code"], details=result.get("details")))
                await self._read_settled.wait()
                if result["status"] == "ended":
                    # Release need not deliver queued data; authoritative end
                    # may follow a prefix deliberately discarded by disposal.
                    # It cannot precede any already committed read, including
                    # a read whose response raced with this release response.
                    if (result["lastSequence"] < self._last_sequence
                        or (self._ended and result["lastSequence"] != self._last_sequence)):
                        await self._runtime._fatal_close("PROTOCOL_ERROR")
                        raise OperationError("PROTOCOL_ERROR", "Inconsistent channel release end")

            self._release_task = self._runtime._track_settlement(dispose())
        return self._release_task

    async def aclose(self) -> None:
        if self._ended and self._release_task is None:
            return
        try:
            await asyncio.shield(self._begin_release())
        except OperationError:
            # Ownership failures cannot be mistaken for successful disposal.
            raise
        if self._cause is not None and not self._exposed:
            self._exposed = True
            raise self._cause

    async def __aenter__(self) -> _Receiver:
        return self

    async def __aexit__(self, exc_type: object, exc: object, tb: object) -> None:
        await self.aclose()


@dataclass(frozen=True, slots=True)
class _Pair:
    send: _Sender
    receive: _Receiver


@dataclass(frozen=True, slots=True)
class _Broadcast:
    send: _Sender
    _runtime: _Runtime
    _reference: str

    async def subscribe(self) -> _Receiver:
        return await self._runtime._subscribe(self)
